internal/redirects/redirect_store.go

```diff
@@ -0,0 +1,209 @@
+		return nil
+	}
+}
+
+func queryRenkuApi(renkuCredentials ServerCredentials, endpoint string) []byte {
+	method := "GET"
+
+	req, err := http.NewRequest(method, fmt.Sprintf("https://%s/api/data%s", renkuCredentials.Host, endpoint), nil)
+	if err != nil {
+		fmt.Printf("Error creating request: %v\n", err)
+		return nil
+	}
+	req.Header.Set("Accept", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		fmt.Printf("Error fetching migrated projects: %v\n", err)
+		return nil
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		fmt.Printf("Unexpected status code: %d\n", resp.StatusCode)
+		return nil
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Printf("Error reading response body: %v\n", err)
+		return nil
+	}
+	return body
+}
+
+func retrieveRedirectForUrl(renkuCredentials ServerCredentials, sourceUrl string) *PlatformRedirectConfig {
+	// Query the Renku API to get the redirect for the given source URL
+	body := queryRenkuApi(renkuCredentials, fmt.Sprintf("/platform/redirects/%s", sourceUrl))
+	if body == nil {
+		return nil
+	}
+
+	var redirectConfig PlatformRedirectConfig
+	if err := json.Unmarshal(body, &redirectConfig); err != nil {
+		fmt.Printf("Error parsing JSON response: %v\n", err)
+		return nil
+	}
+
+	return &redirectConfig
+}
+
+func (rs *RedirectStore) GetRedirectEntry(key string) (*RedirectStoreRedirectEntry, error) {
+	if rs == nil {
+		return nil, fmt.Errorf("redirect store is not initialized")
+	}
+
+	entry, ok := rs.RedirectMap[key]
+	if ok && entry.UpdatedAt.Add(rs.EntryTtl).After(time.Now()) {
+		return &entry, nil
+	}
+
+	rs.redirectMapMutex.Lock()
+	defer rs.redirectMapMutex.Unlock()
+	// Re-check after acquiring the lock, since it might have been updated meanwhile
+	entry, ok = rs.RedirectMap[key]
+	if !ok || entry.UpdatedAt.Add(rs.EntryTtl).Before(time.Now()) {
+		updatedEntry := retrieveRedirectForUrl(ServerCredentials{
+			Host: rs.Config.RenkuBaseURL.Host,
+		}, key)
+		if updatedEntry == nil {
+			// No entry, this is fine
+			return nil, nil
+		}
+		entry = RedirectStoreRedirectEntry{
+			SourceUrl: updatedEntry.SourceUrl,
+			TargetUrl: &updatedEntry.TargetUrl,
+			UpdatedAt: time.Now(),
+		}
+		rs.RedirectMap[key] = entry
+	}
+	return &entry, nil
+}
+
+func (rs *RedirectStore) Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			url := c.Request().URL
+			if url == nil {
+				return next(c)
+			}
+
+			path := url.Path
+			if path == "" {
+				// This should not happen because the path should start with PathPrefix
+				return next(c)
+			}
+			if !strings.HasPrefix(path, rs.PathPrefix) {
+				return next(c)
+			}
+
+			urlToCheck := strings.TrimPrefix(path, rs.PathPrefix)
+			// TODO: Check for a `/-/` in the path and remove it and anything that follows (links to sub-pages of a project)
+			urlToCheck = fmt.Sprintf("https://%s/%s", rs.redirectedHost, urlToCheck)
```

Comment:
suggestion: we should probably check for both the `http` and `https` url in the data service. I don think the data service is picky about this. So you could have one or the other.